NetKet/GroundState/exact_diagonalization.hpp

```diff
@@ -23,52 +23,98 @@
 #include "Hamiltonian/MatrixWrapper/matrix_wrapper.hpp"
 
 namespace netket {
+  
+  namespace detail {
+    template <class matrix_t, class iter_t, class random_t>
+    std::vector<double>eigenvalues_lanczos_run(const matrix_t& matrix,
+					       const random_t& random_gen,
+					       iter_t& iter) {
+      using complex = std::complex<double>;
+      using vectorspace_t = ietl::vectorspace<complex>;
+      using lanczos_t = ietl::lanczos<matrix_t, vectorspace_t>;
+      size_t dimension = matrix.GetDimension();
+      vectorspace_t ietl_vecspace(dimension);
+      lanczos_t lanczos(matrix, ietl_vecspace);
+      lanczos.calculate_eigenvalues(iter, random_gen);
+      if (iter.error_code() == 1)
+	printf("Warning: Lanczos algorithm did NOT converge in %i steps!\n",
+	       iter.max_iterations());      
+      return lanczos.eigenvalues();
+    }
+  }
+  
+  std::vector<double> eigenvalues_lanczos(const Hamiltonian &hamiltonian,
+					  bool matrix_free=false,
+					  int first_n = 1,
+					  int max_iter = 1000, int seed=42,
+					  double precision = 1e-14) {
+    using normal_dist_t = std::uniform_real_distribution<double>;
+    using random_t = ietl::random_generator<std::mt19937, normal_dist_t>;
+    using iter_t = ietl::lanczos_iteration_nlowest<double>;
+
+    normal_dist_t dist(-1., 1.);
+    random_t random_gen(dist, seed);
+
+    // Converge the first_n eigenvalues to precision
+    iter_t iter(max_iter, first_n, sqrt(precision), sqrt(precision));
+
+    std::vector<double> eigs_lanczos;
+  
+    if (matrix_free) { // Matrix-free computation
```

Comment:
This part has some code duplication. How about doing something like this at an appropriate point in the code:
```cpp
auto matrix = ConstructMatrixWrapper(pars);
// ...
detail::eigenvalue_lanczos_run(..., *matrix)
```
The `ConstructMatrixWrapper` function returns `unique_ptr<AbstractMatrixWrapper>` and any child of `AbstractMatrixWrapper` will work with your version of `ietl::mult` (which is the idea behind having an abstract base class for the matrix wrappers).
Users can specify the desired wrapper via e.g. `"MatrixWrapper": "Sparse"` or  `"MatrixWrapper": "Direct"` in the JSON, just like for the imaginary time propagation.